### PR TITLE
Simplified and fixed estimate computation.

### PR DESCRIPTION
**Pull request by StephanEwen**
Thsi patch greatly simplifies the estimate computation. To that end, it also simplifies the compiler hints.

The prior version was more expressive, but was never used because it required the programmer to ahve detailed knowledge about the program (like per column group cardinalities).

In the long run, we need a clear plan how to best expose hints. Until then, I think this makes it much simpler to use and it fixed a buch of bugs on the way.

### CODE REVIEW

**Comment by StephanEwen** [comment]
Any objections to merging this?

@fhueske You have worked on that part before as well, do you see any complications with this?

**Comment by fhueske** [comment]
I'll have a look at it.

File: stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/record/relational/WebLogAnalysis.java [hunk 1]
```diff
@@ -279,25 +278,19 @@ public Plan getPlan(String... args) {
...
-		filterDocs.getCompilerHints().setAvgBytesPerRecord(60);
-		filterDocs.getCompilerHints().setAvgNumRecordsPerDistinctFields(new FieldSet(new int[]{0}), 1);
+		filterDocs.getCompilerHints().setFilterFactor(0.15f);
 
 		// Create MapOperator for filtering the entries from the ranks relation
 		MapOperator filterRanks = MapOperator.builder(new FilterRanks())
 			.input(ranks)
 			.name("Filter Ranks")
 			.build();
-		filterRanks.getCompilerHints().setAvgRecordsEmittedPerStubCall(0.25f);
```
Comment (fhueske) [inline stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/record/relational/WebLogAnalysis.java:291]: add filterFactor hint for consistency with Scala example.

File: stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/record/relational/WebLogAnalysis.java [hunk 2]
```diff
@@ -279,25 +278,19 @@ public Plan getPlan(String... args) {
...
-		filterRanks.getCompilerHints().setAvgNumRecordsPerDistinctFields(new FieldSet(new int[]{0}), 1);
 
 		// Create MapOperator for filtering the entries from the visits relation
 		MapOperator filterVisits = MapOperator.builder(new FilterVisits())
 			.input(visits)
 			.name("Filter Visits")
 			.build();
-		filterVisits.getCompilerHints().setAvgBytesPerRecord(60);
-		filterVisits.getCompilerHints().setAvgRecordsEmittedPerStubCall(0.2f);
```
Comment (fhueske) [inline stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/record/relational/WebLogAnalysis.java:300]: add filterFactor hint for consistency with Scala example

File: stratosphere-core/src/main/java/eu/stratosphere/api/common/operators/CompilerHints.java [hunk 3]
```diff
@@ -13,47 +13,26 @@
...
- * the function encapsulated in the contract. For example, if a function declares a selectivity of 0.5, it declares that
- * it produces roughly half as many key/value pairs as it consumes. Its default value is <code>-1.0</code>, indicating
- * that the selectivity is unknown.</li>
- * <li>The average number of values per distinct key. <code>-1.0</code> by default, indicating that this value is
- * unknown.</li>
- * <li>The average number of bytes per record. <code>-1.0</code> by default, indicating that this value is unknown.</li>
- * </ul>
+ * A class encapsulating compiler hints describing the behavior of the user function.
+ * If set, the optimizer will use them to estimate the sizes of the intermediate results.
+ * Not that these values are optional hints, the optimizer will always generate a valid plan without
```
Comment (fhueske) [inline stratosphere-core/src/main/java/eu/stratosphere/api/common/operators/CompilerHints.java]: Not -> Note

File: stratosphere-core/src/main/java/eu/stratosphere/api/common/operators/CompilerHints.java [hunk 4]
```diff
@@ -13,47 +13,26 @@
...
-
-	private float avgBytesPerRecord = -1.0f;
+	private long outputSize = -1;
 	
-	private Map<FieldSet, Long> distinctCounts = new HashMap<FieldSet, Long>();
-
-	private Map<FieldSet, Float> avgNumRecordsPerDistinctFields = new HashMap<FieldSet, Float>();
+	private long outputCardinality = -1;
+	
+	private float avgBytesPerOutputRecord = -1.0f;
```
Comment (fhueske) [inline stratosphere-core/src/main/java/eu/stratosphere/api/common/operators/CompilerHints.java]: rename to `avgOutputRecordSize` to be consistent with `outputSize` hint?

File: stratosphere-compiler/src/main/java/eu/stratosphere/compiler/plandump/PlanJSONDumpGenerator.java [hunk 5]
```diff
@@ -547,35 +536,17 @@ private void visit(DumpableNode<?> node, PrintWriter writer, boolean first) {
...
-			} else {
-				valuesKey = "(none)";
-			}
-			addProperty(writer, "Avg. Values/Distinct fields", valuesKey, false);
-			
-			addProperty(writer, "Avg. Width (bytes)", hints.getAvgBytesPerRecord() == defaults
-				.getAvgBytesPerRecord() ? "(none)" : String.valueOf(hints.getAvgBytesPerRecord()), false);
+			addProperty(writer, "Output Size", size, true);
+			addProperty(writer, "Output Cardinality", card, false);
+			addProperty(writer, "Output avg. width (bytes)", width, false);
```
Comment (fhueske) [inline stratosphere-compiler/src/main/java/eu/stratosphere/compiler/plandump/PlanJSONDumpGenerator.java]: Avg. Output Record Size (bytes)
In any case consistent with the optimizer hint which do not use "width" right now.

File: stratosphere-compiler/src/main/java/eu/stratosphere/compiler/plandump/PlanJSONDumpGenerator.java [hunk 6]
```diff
@@ -547,35 +536,17 @@ private void visit(DumpableNode<?> node, PrintWriter writer, boolean first) {
...
-					valuesKey += "[" + entry.getKey().toString() + "->" + entry.getValue() + "]"; 
-				}
-			} else {
-				valuesKey = "(none)";
-			}
-			addProperty(writer, "Avg. Values/Distinct fields", valuesKey, false);
-			
-			addProperty(writer, "Avg. Width (bytes)", hints.getAvgBytesPerRecord() == defaults
-				.getAvgBytesPerRecord() ? "(none)" : String.valueOf(hints.getAvgBytesPerRecord()), false);
+			addProperty(writer, "Output Size", size, true);
```
Comment (fhueske) [inline stratosphere-compiler/src/main/java/eu/stratosphere/compiler/plandump/PlanJSONDumpGenerator.java]: Output Size (bytes)

File: stratosphere-compiler/src/main/java/eu/stratosphere/compiler/dag/MapNode.java [hunk 7]
```diff
@@ -49,20 +45,17 @@ public String getName() {
...
-			return -1;
-	}
-
 	@Override
 	protected List<OperatorDescriptorSingle> getPossibleProperties() {
 		return Collections.<OperatorDescriptorSingle>singletonList(new MapDescriptor());
 	}
+
+	/**
+	 * Computes the estimates for the Map operator. Map takes one value and transforms it into another value.
```
Comment (fhueske) [inline stratosphere-compiler/src/main/java/eu/stratosphere/compiler/dag/MapNode.java]: Comment assumes Map is only used as transformation and not as a filter or "splitter".
I would rather say, that input cardinality is the default assumption.

**Comment by fhueske** [comment]
The PR reduces and eases the set of compiler hints.
It also significantly cleans up the estimates computation in the compiler.

We lose some expressiveness through the removed compiler hints, but these were hard to use in any case.
I have some cosmetic remarks, but nothing curcial.

After that it's good to merge, IMHO.

**Comment by fhueske** [comment]
Just made a PR on branch of this PR to address my comments.

**Comment by fhueske** [comment]
@StephanEwen Why didn't you merge it into master?

**Comment by StephanEwen** [comment]
Tests are still running... will merge as soon as they turn green.

**Pull request closed without merging**